torch/dispatch/dispatcher.py

```diff
@@ -0,0 +1,95 @@
+import torch
+import torch._C as _C
+from torch._C import DispatchKey, DispatchKeySet, ExcludeDispatchKeyGuard  # type: ignore[attr-defined]
+from torch.utils._pytree import tree_flatten
+from torch.overrides import handle_torch_function, has_torch_function
+
+SUPPORTED_KEYS = {
+    DispatchKey.CPU,
+    DispatchKey.BackendSelect,
+    DispatchKey.ADInplaceOrView,
+    DispatchKey.AutogradCPU,
+    DispatchKey.Python,
+    DispatchKey.PythonTLSSnapshot,
+}
+
+"""
+This is a dispatcher (in Python)
+- You can define new operations (in Python) without schemas
+- It interfaces with the PyTorch dispatcher
+"""
+
+class PyDispatcher:
+    def __init__(self):
+        self.current_dispatching_op = None
+        self.already_dispatched_keys = None
+        self.current_key = None
```

Comment:
This state isn't in the C++ dispatcher, looks a bit suspicious

Reply:
I added current_key to the dispatcher to add support for "re-entrant" dispatching.

Reply:
There's more work this took, but I removed it.